Implement NcInfo guess() function

Currently in line 176 of [https://github.com/georust/netcdf/blob/master/netcdf-sys/build.rs](https://github.com/georust/netcdf/blob/master/netcdf-sys/build.rs) it says:

```
impl NcInfo {
    fn guess() -> Self {
        todo!()
    }
```

Which caused me an error when doing `cargo run` on my project after just doing `cargo add netcdf`, installing HDF5 (since it worked for somebody else and was causing me issues) and adding `use netcdf;` to my project.

Any help is much appreciated.

The error message I got is
```
Compiling netcdf-sys v0.8.2
error: failed to run custom build command for `netcdf-sys v0.8.2`

Caused by:
  process didn't exit successfully: `my_project\target\debug\build\netcdf-sys-3c2aaa4a2a4f361b\build-script-build` (exit code: 101)
  --- stdout
  cargo::rerun-if-changed=build.rs
  cargo::rerun-if-env-changed=NETCDF_DIR

  --- stderr

  thread 'main' panicked at C:\Users\G0rocks\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\netcdf-sys-0.8.2\build.rs:176:9:
  not yet implemented
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

I installed cmake and verified it with `cmake --version` and got `cmake version 4.0.2`.
If you need cmake to use the crate then maybe that should also be added to the main page of the docs?

Now I get this error:

```
  --- stderr
  CMake Error at CMakeLists.txt:2 (project):
    No CMAKE_C_COMPILER could be found.

  thread 'main' panicked at C:\Users\G0rocks\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\cmake-0.1.54\src\lib.rs:1119:5:

  command did not execute successfully, got: exit code: 1

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

--- a/.typos.toml
+++ b/.typos.toml
@@ -5,4 +5,5 @@ extend-exclude = ["netcdf-src/source", "netcdf-sys"]
 extend-ignore-identifiers-re = [
   "Dout",
   "is_ambigous", # Remove when deprecated item is removed
+  "typ",
 ]

--- a/README.md
+++ b/README.md
@@ -31,7 +31,9 @@ All variable data is read into a contiguous buffer, or into an [ndarray](https:/
 
 ## Building
 
-This crate depends on the library [`netcdf-c`](https://www.unidata.ucar.edu/netcdf/) which must be installed on the machine, along with libraries such as `hdf5`. An alternative to the system libraries is the use of the `static` feature of this crate (`cargo add netcdf --features static`), which compiles `libnetcdf` from source. The `static` feature requires `cmake`, a `c++` compiler and more to be installed on the build machine.
+This crate depends on the C library [`netcdf-c`](https://www.unidata.ucar.edu/netcdf/) which must be installed on the machine, along with libraries such as `hdf5` which `netcdf-c` depends on. For some platforms you might have to set the environment variable `NETCDF_DIR` which needs to point at the installation of `netcdf-c` you wish to use.
+
+An alternative to using the system libraries is to enable `static` feature of this crate (`cargo add netcdf --features static`), which compiles `libnetcdf` from source. The `static` feature requires `cmake`, a `c++` compiler and more to be installed on the build machine.
 
 The crate is built on several platforms using github actions, and is currently known to build form from source on all major platforms (linux, macos, windows (gnu+msvc)), and through the package installers `conda` and `apt`. Please see the github workflows for tips on how to install `netcdf`.
 

--- a/README.md
+++ b/README.md
@@ -31,13 +31,12 @@ All variable data is read into a contiguous buffer, or into an [ndarray](https:/
 
 ## Building
 
-This crate depends on `libnetcdf`, but a static build from source is also supported, which can be enabled using the `static` feature.
+This crate depends on the library [`netcdf-c`](https://www.unidata.ucar.edu/netcdf/) which must be installed on the machine, along with libraries such as `hdf5`. An alternative to the system libraries is the use of the `static` feature of this crate (`cargo add netcdf --features static`), which compiles `libnetcdf` from source. The `static` feature requires `cmake`, a `c++` compiler and more to be installed on the build machine.
 
-The crate is built on several platforms using github actions, and is currently known to build form from source on all major platforms (linux, macos, windows (gnu+msvc)), and through the package installers `conda` and `apt`.
+The crate is built on several platforms using github actions, and is currently known to build form from source on all major platforms (linux, macos, windows (gnu+msvc)), and through the package installers `conda` and `apt`. Please see the github workflows for tips on how to install `netcdf`.
 
-If during compilation there is an error in building the `hdf5` crate, consider using the `static` feature which will include a compatible version of both `netcdf` and `hdf5`. This is likely to be an issue [upstream](https://github.com/aldanor/hdf5-rust/issues/262).
 
-### Building without `libnetcdf` or building statically
+### Building `libnetcdf` statically
 1. `git clone https://github.com/georust/netcdf`
 2. `git submodule update --init --recursive`
 3. `cargo build --features static`

--- a/netcdf-sys/build.rs
+++ b/netcdf-sys/build.rs
@@ -249,8 +249,25 @@ fn main() {
             .or_else(|| std::env::var_os("NetCDF_DIR"))
             .map(PathBuf::from);
 
+        #[cfg(windows)]
+        let nc_dir = nc_dir.map(|d| d.join("Library"));
+
         info = if let Some(nc_dir) = nc_dir.as_ref() {
-            NcInfo::gather_from_ncconfig(Some(nc_dir)).unwrap_or_else(|| NcInfo::from_path(nc_dir))
+            #[allow(unused_mut)]
+            let mut info = NcInfo::gather_from_ncconfig(Some(nc_dir))
+                .unwrap_or_else(|| NcInfo::from_path(nc_dir));
+            #[cfg(windows)]
+            {
+                // Conda requires Library prefix
+                if info.is_none() {
+                    let nc_dir = nc_dir.map(|d| d.join("Library")).unwrap();
+
+                    info = NcInfo::gather_from_ncconfig(Some(nc_dir))
+                        .unwrap_or_else(|| NcInfo::from_path(nc_dir));
+                }
+            }
+
+            info
         } else {
             NcInfo::gather_from_ncconfig(None).unwrap_or_else(||
                 panic!("A system version of libnetcdf could not be found. Consider installing to some default location, use the environment variable NETCDF_DIR (remember to restart the shell), or prefer building the static version of libnetcdf by enabling the `static` feature on `netcdf-sys` or `netcdf`")

--- a/netcdf-sys/build.rs
+++ b/netcdf-sys/build.rs
@@ -172,13 +172,18 @@ fn from_utf8_to_trimmed_string(bytes: &[u8]) -> String {
 }
 
 impl NcInfo {
-    fn from_path(path: &Path) -> Self {
-        Self {
+    fn from_path(path: &Path) -> Option<Self> {
+        let includedir = path.join("include");
+        if !includedir.exists() {
+            return None;
+        }
+
+        Some(Self {
             version: None,
             includedir: path.join("include"),
             libdir: path.join("lib"),
             libname: "netcdf".to_owned(),
-        }
+        })
     }
     fn gather_from_ncconfig(search_path: Option<&Path>) -> Option<Self> {
         let path = if let Some(search_path) = search_path {
@@ -229,6 +234,40 @@ fn _check_consistent_version_linked() {
     todo!()
 }
 
+fn determine_ncinfo() -> NcInfo {
+    let nc_dir = std::env::var_os("NETCDF_DIR")
+        .or_else(|| std::env::var_os("NetCDF_DIR"))
+        .map(PathBuf::from);
+
+    if let Some(nc_dir) = nc_dir.as_ref() {
+        #[allow(unused_mut)]
+        if let Some(info) = NcInfo::gather_from_ncconfig(Some(nc_dir)) {
+            return info;
+        }
+        if let Some(info) = NcInfo::from_path(nc_dir) {
+            return info;
+        }
+        #[cfg(windows)]
+        {
+            // Conda requires Library prefix
+            let nc_dir = nc_dir.join("Library");
+            if let Some(info) = NcInfo::gather_from_ncconfig(Some(&nc_dir)) {
+                return info;
+            }
+            if let Some(info) = NcInfo::from_path(&nc_dir) {
+                return info;
+            }
+        }
+    }
+
+    if let Some(info) = NcInfo::gather_from_ncconfig(None) {
+        return info;
+    }
+
+    panic!("A system version of libnetcdf could not be found. Consider installing to some default location, use the environment variable NETCDF_DIR (remember to restart the shell), or prefer building the static version of libnetcdf by enabling the `static` feature on `netcdf-sys` or `netcdf`"
+            );
+}
+
 fn main() {
     println!("cargo::rerun-if-changed=build.rs");
 
@@ -237,42 +276,19 @@ fn main() {
         let netcdf_lib = std::env::var("DEP_NETCDFSRC_LIB").unwrap();
         let netcdf_path = PathBuf::from(std::env::var_os("DEP_NETCDFSRC_SEARCH").unwrap());
 
-        info = NcInfo::gather_from_ncconfig(Some(&netcdf_path.join("..")))
-            .unwrap_or_else(|| NcInfo::from_path(&netcdf_path.join("..")));
+        let mut linfo = NcInfo::gather_from_ncconfig(Some(&netcdf_path.join("..")));
+        if linfo.is_none() {
+            linfo = NcInfo::from_path(&netcdf_path.join(".."))
+        }
+
+        info = linfo.expect("The library path should be determined by this point");
 
         println!("cargo::rustc-link-search=native={}", netcdf_path.display());
         println!("cargo::rustc-link-lib=static={netcdf_lib}");
     } else {
         println!("cargo::rerun-if-env-changed=NETCDF_DIR");
 
-        let nc_dir = std::env::var_os("NETCDF_DIR")
-            .or_else(|| std::env::var_os("NetCDF_DIR"))
-            .map(PathBuf::from);
-
-        #[cfg(windows)]
-        let nc_dir = nc_dir.map(|d| d.join("Library"));
-
-        info = if let Some(nc_dir) = nc_dir.as_ref() {
-            #[allow(unused_mut)]
-            let mut info = NcInfo::gather_from_ncconfig(Some(nc_dir))
-                .unwrap_or_else(|| NcInfo::from_path(nc_dir));
-            #[cfg(windows)]
-            {
-                // Conda requires Library prefix
-                if info.is_none() {
-                    let nc_dir = nc_dir.map(|d| d.join("Library")).unwrap();
-
-                    info = NcInfo::gather_from_ncconfig(Some(nc_dir))
-                        .unwrap_or_else(|| NcInfo::from_path(nc_dir));
-                }
-            }
-
-            info
-        } else {
-            NcInfo::gather_from_ncconfig(None).unwrap_or_else(||
-                panic!("A system version of libnetcdf could not be found. Consider installing to some default location, use the environment variable NETCDF_DIR (remember to restart the shell), or prefer building the static version of libnetcdf by enabling the `static` feature on `netcdf-sys` or `netcdf`")
-            )
-        };
+        info = determine_ncinfo();
 
         println!("cargo::rustc-link-search={}", info.libdir.display());
         println!("cargo::rustc-link-lib={}", &info.libname);

--- a/netcdf-sys/build.rs
+++ b/netcdf-sys/build.rs
@@ -172,9 +172,6 @@ fn from_utf8_to_trimmed_string(bytes: &[u8]) -> String {
 }
 
 impl NcInfo {
-    fn guess() -> Self {
-        todo!()
-    }
     fn from_path(path: &Path) -> Self {
         Self {
             version: None,
@@ -258,7 +255,9 @@ fn main() {
         info = if let Some(nc_dir) = nc_dir.as_ref() {
             NcInfo::gather_from_ncconfig(Some(nc_dir)).unwrap_or_else(|| NcInfo::from_path(nc_dir))
         } else {
-            NcInfo::gather_from_ncconfig(None).unwrap_or_else(NcInfo::guess)
+            NcInfo::gather_from_ncconfig(None).unwrap_or_else(||
+                panic!("A system version of libnetcdf could not be found. Consider installing to some default location, use NETCDF_DIR, or prefer the static version of libnetcdf by setting the `static` feature on `netcdf-sys`")
+            )
         };
 
         println!("cargo::rustc-link-search={}", info.libdir.display());

--- a/netcdf-sys/build.rs
+++ b/netcdf-sys/build.rs
@@ -249,14 +249,11 @@ fn main() {
             .or_else(|| std::env::var_os("NetCDF_DIR"))
             .map(PathBuf::from);
 
-        #[cfg(windows)]
-        let nc_dir = nc_dir.map(|d| d.join("Library"));
-
         info = if let Some(nc_dir) = nc_dir.as_ref() {
             NcInfo::gather_from_ncconfig(Some(nc_dir)).unwrap_or_else(|| NcInfo::from_path(nc_dir))
         } else {
             NcInfo::gather_from_ncconfig(None).unwrap_or_else(||
-                panic!("A system version of libnetcdf could not be found. Consider installing to some default location, use NETCDF_DIR, or prefer the static version of libnetcdf by setting the `static` feature on `netcdf-sys`")
+                panic!("A system version of libnetcdf could not be found. Consider installing to some default location, use the environment variable NETCDF_DIR (remember to restart the shell), or prefer building the static version of libnetcdf by enabling the `static` feature on `netcdf-sys` or `netcdf`")
             )
         };
 

--- a/netcdf/src/lib.rs
+++ b/netcdf/src/lib.rs
@@ -14,7 +14,14 @@
 //!
 //! For more information see:
 //! * [The official introduction to `netCDF`](https://docs.unidata.ucar.edu/nug/current/netcdf_introduction.html)
-//! * [The `netCDF-c` repository](https://github.com/Unidata/netcdf-c)
+//! * [The `NetCDF-c` repository](https://github.com/Unidata/netcdf-c)
+//!
+//! # Installing netcdf-c
+//!
+//! This crate depends on [Unidata NetCDF-c](https://github.com/Unidata/netcdf-c) and dependencies
+//! of this library (such as hdf5).
+//! An alternative to the system libraries is to use the bundled sources of netcdf by using the `static` feature of this crate. This will require build utilities such as `cmake`, `c++` compiler and more.
+//!
 //!
 //! # Examples
 //!

--- a/netcdf/src/variable.rs
+++ b/netcdf/src/variable.rs
@@ -19,7 +19,7 @@ use crate::utils::{checked_with_lock, with_lock};
 ///
 /// This type is used for retrieving data from a variable.
 /// Metadata on the `netCDF`-level can be retrieved using e.g.
-/// [`fill_value`](Self::fill_value), [`endinanness`](Self::endianness).
+/// [`fill_value`](Self::fill_value), [`endianness`](Self::endianness).
 /// Use [`attributes`](Self::attribute) to get additional metadata assigned
 /// by the data producer. This crate will not apply any of the transformations
 /// given by such attributes (e.g. `add_offset` and `scale_factor` are NOT considered).

